Content.Server/Database/ServerDbManager.cs

```diff
@@ -86,131 +48,360 @@ public void Init()
                 builder.AddProvider(_msLogProvider);
             });
 
+            ServerDbContext = CreateDbContext();
+
+            ServerDbContext.Database.Migrate();
+        }
+
+        protected virtual ServerDbContext CreateDbContext()
+        {
             var engine = _cfg.GetCVar(CCVars.DatabaseEngine).ToLower();
             switch (engine)
             {
                 case "sqlite":
                     var options = CreateSqliteOptions();
-                    _db = new ServerDbSqlite(options);
-                    break;
+                    return new SqliteServerDbContext(options);
                 case "postgres":
                     options = CreatePostgresOptions();
-                    _db = new ServerDbPostgres(options);
-                    break;
+                    return new PostgresServerDbContext(options);
                 default:
                     throw new InvalidDataException("Unknown database engine {engine}.");
             }
         }
 
-        public Task<PlayerPreferences> InitPrefsAsync(NetUserId userId, ICharacterProfile defaultProfile)
+        public async Task<PlayerPreferences> InitPrefsAsync(NetUserId userId, ICharacterProfile defaultProfile)
         {
-            return _db.InitPrefsAsync(userId, defaultProfile);
+
+            var profile = new Profile
+            {
+                Slot = 0
+            };
+            ((HumanoidCharacterProfile) defaultProfile).ConvertProfile(profile);
+
+            var prefs = await ServerDbContext.Preferences
+                .Where(p => p.UserId == userId)
+                .SingleOrDefaultAsync();
+
+            // To explain: The code later will return null if it encounters preferences and no profiles.
+            // This is because the preferences are corrupt, and the DB code doesn't have access to the function to create a new profile.
+            // In this case, a Preference object already exists but is safe to reuse.
+            if (prefs is null)
```

Comment:
Isn't this just always going to throw a warning the first time somebody joins the server ??

Reply:
This is my fault. I was trying to catch cases where no profiles were available.